### DISCUSSION TITLE
creating-a-source-generator-part-7-solving-the-source-generator-marker-attribute-problem-part1/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#366] [Announcements]
# creating-a-source-generator-part-7-solving-the-source-generator-marker-attribute-problem-part1/

In this post I describe a problem I've been wrestling with around source generators: where to put the 'marker attributes' that drive the generator

https://andrewlock.net/creating-a-source-generator-part-7-solving-the-source-generator-marker-attribute-problem-part1/

[a comment by an automated account is not shown]

**Comment by BlinD-HuNTeR** [comment]
The preferred approach is indeed to use the PostInitialization callback, with `internal` visibility. The usage of `InternalsVisibleTo` is very rare, and even when it's used, the compiler will always choose types defined in the current project over types from external references, generating only a warning, and not an error.

>Personnaly, I would use the source generator to add the attribute to the compilation, and use the `CompilerVisibleProperty` MSBuild ItemGroup to add an "option" which can be set on the csproj to disable writing the attribute code as a workaround. It will be required only for very specific scenarios, and instructions will be easy to share on the project documentation.

There's no real need to make it configurable, it will always work correctly with `internal` visibility, and most of the time, no warning will be produced. If someone ever gets a warning due to `InternalsVisibleTo`, they can always suppress it at their project, which is easier to do than messing with the generator's options (VS can even do that automatically).

**Reply by Serdan** [reply]
`EmbeddedAttribute` will hide an internal type even with `InternalsVisibleTo`

https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.cookbook.md#put-microsoftcodeanalysisembeddedattribute-on-generated-marker-types

**Comment by AndersKehlet** [comment]
From 4.14 the "correct" way to do this will be to call `context.RegisterPostInitializationOutput(ctx => ctx.AddEmbeddedAttributeDefinition());`

You can already add EmbeddedAttribute to your internal types, but you need to declare it yourself and add it to the compilation along with the marker attributes. Annotated types are not visible to other assemblies, even with `InternalsVisibleTo`.

https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.cookbook.md#put-microsoftcodeanalysisembeddedattribute-on-generated-marker-types

**Reply by AndersKehlet** [reply]
From my own project:

```csharp
using System;

namespace Microsoft.CodeAnalysis
{
    /// <summary>
    /// Hides internal types even with InternalsVisibleTo.
    /// IncrementalGeneratorPostInitializationContext.AddEmbeddedAttributeDefinition() will be available from Roslyn 4.14
    /// This file should be removed at that time. 
    /// </summary>
    internal sealed class EmbeddedAttribute : Attribute;
}
```
